core/src/main/java/org/apache/iceberg/rest/responses/ErrorResponseParser.java

```diff
@@ -0,0 +1,101 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.rest.responses;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.io.UncheckedIOException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.JsonUtil;
+
+public class ErrorResponseParser {
+
+  private ErrorResponseParser() {
+  }
+
+  private static final String ERROR = "error";
+  private static final String MESSAGE = "message";
+  private static final String TYPE = "type";
+  private static final String CODE = "code";
+
+  public static String toJson(ErrorResponse errorResponse) {
+    return toJson(errorResponse, false);
+  }
+
+  public static String toJson(ErrorResponse errorResponse, boolean pretty) {
+    try {
+      StringWriter writer = new StringWriter();
+      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
+      if (pretty) {
+        generator.useDefaultPrettyPrinter();
+      }
+      toJson(errorResponse, generator);
+      generator.flush();
+      return writer.toString();
+    } catch (IOException e) {
+      throw new UncheckedIOException(String.format("Failed to write error response json for: %s", errorResponse), e);
+    }
+  }
+
+  public static void toJson(ErrorResponse errorResponse, JsonGenerator generator) throws IOException {
+    generator.writeStartObject();
+
+    generator.writeObjectFieldStart(ERROR);
+
+    generator.writeStringField(MESSAGE, errorResponse.message());
+    generator.writeStringField(TYPE, errorResponse.type());
+    generator.writeNumberField(CODE, errorResponse.code());
+
+    generator.writeEndObject();
+
+    generator.writeEndObject();
+  }
+
+  /**
+   * Read ErrorResponse from a JSON string.
+   *
+   * @param json a JSON string of an ErrorResponse
+   * @return an ErrorResponse object
+   */
+  public static ErrorResponse fromJson(String json) {
+    try {
+      return fromJson(JsonUtil.mapper().readValue(json, JsonNode.class));
+    } catch (IOException e) {
+      throw new UncheckedIOException("Failed to read JSON string: " + json, e);
+    }
+  }
+
+  public static ErrorResponse fromJson(JsonNode jsonNode) {
+    Preconditions.checkArgument(jsonNode != null && jsonNode.isObject() && jsonNode.has(ERROR),
+        "Cannot parse missing field: error");
```

Comment:
This mixes two different errors. The first is that the JSON node has the wrong type and can't be the error response. The second is that it doesn't have an "error" field.

Reply:
Updted to two checks.